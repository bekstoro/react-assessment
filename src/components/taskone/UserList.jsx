import React, { useCallback, useEffect, useState } from 'react';
import { Row, UserInfo, Users } from '@components/styled-components/UserList';
import { debounce } from '@utils/debounce';

export default function UserList() {
  const [data, setData] = useState([]);
  const [filter, setFilter] = useState('');
  const [value, setValue] = useState('');

  const fetchUsers = useCallback(async () => {
    fetch(`https://jsonplaceholder.typicode.com/users${filter ? `?username=${encodeURIComponent(filter)}` : ''}`)
      .then(async (response) => {
        const res = await response.json();
        setData(res);
      });
  }, [filter]);

  useEffect(() => {
    fetchUsers();
  }, [fetchUsers, filter]);

  const handleFilter = (e) => {
    setValue(e.target.value);
    const debounceFn = debounce(() => {
      setFilter(e.target.value);
    });

    debounceFn();
  };

  return (
    <div>
      <div>
        Filter:
        <input
          type="text"
          onChange={handleFilter}
          value={value}
          placeholder="Enter username"
        />
      </div>
      <Users>
        {data.map((user) => (
          <Row key={user.id}>
            <UserInfo>
              <span>{`Name: ${user.name}`}</span>
              <span>{`Username: ${user.username}`}</span>
            </UserInfo>
            <div>
              <div>
                <span>{user.address?.street}</span>
                <span>{user.address?.suite}</span>
                <span>{user.address?.city}</span>
                <span>{user.address?.zipcode}</span>
              </div>
              <div>
                <span>{user.email}</span>
                <span>{user.phone}</span>
              </div>
            </div>
          </Row>
        ))}
      </Users>
    </div>
  );
}
