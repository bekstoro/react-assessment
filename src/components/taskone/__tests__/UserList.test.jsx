import React from 'react';
import { render } from '@testing-library/react';

import UserList from '@components/taskone/UserList';

describe('UserList', () => {
  it('UserList renders correctly', () => {
    const { container } = render(<UserList />);
    expect(container).toMatchSnapshot();
  });
});
