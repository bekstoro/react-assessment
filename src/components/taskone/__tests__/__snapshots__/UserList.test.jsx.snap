// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList UserList renders correctly 1`] = `
<div>
  <div>
    <div>
      Filter:
      <input
        placeholder="Enter username"
        type="text"
        value=""
      />
    </div>
    <div
      class="UserList__Users-g96ac2-2 fpNOEP"
    />
  </div>
</div>
`;
